src/GraphQL/Types/Composite/ListGraphType.cs

```diff
@@ -27,23 +16,48 @@ public ListGraphType(IGraphType type)
             ResolvedType = type;
         }
 
-        /// <inheritdoc cref="ListGraphType(IGraphType)"/>
-        protected ListGraphType(Type type)
-        {
-            Type = type;
-        }
-
         /// <summary>
         /// Returns the .NET type of the inner (wrapped) graph type.
         /// </summary>
-        public Type Type { get; private set; }
+        public virtual Type Type => null;
+
+        private IGraphType _resolvedType;
 
         /// <summary>
         /// Gets or sets the instance of the inner (wrapped) graph type.
         /// </summary>
-        public IGraphType ResolvedType { get; set; }
+        public IGraphType ResolvedType
+        {
+            get => _resolvedType;
+            set
+            {
+                if (value != null && Type != null && !Type.IsAssignableFrom(value.GetType()))
+                    throw new InvalidOperationException($"Type '{Type.Name}' should be assignable from ResolvedType '{value.Name}'.");
+
+                _resolvedType = value;
+            }
+        }
 
         /// <inheritdoc/>
         public override string ToString() => $"[{ResolvedType}]";
     }
+
+    /// <inheritdoc cref="ListGraphType"/>
+    public class ListGraphType<T> : ListGraphType
+        where T : IGraphType
+    {
+        private readonly Type _type;
+
+        /// <summary>
+        /// Initializes a new instance for the specified inner graph type.
+        /// </summary>
+        public ListGraphType()
+            : base(null)
+        {
+            _type = typeof(T);
+        }
+
+        /// <inheritdoc/>
+        public override Type Type => _type;
```

Comment:
`Type` is only used when the schema is built, correct?  There should be no reason to cache the value.
```suggestion
        public override Type Type => typeof(T);
```

Reply:
🤦